### PR TITLE
Integration of consensus node with network stack

### PR DESCRIPTION
**Pull request by coot**

### CODE REVIEW

**Comment by coot** [comment]
This PR is based on #582, which should be merged first.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 1]
```diff
@@ -284,40 +289,93 @@ initInternalState NodeParams {..} = do
...
-
-        nrBlockFetchServer :: BlockFetchServer blk m ()
-        nrBlockFetchServer =
-          blockFetchServer (tracePrefix "BFServer" Nothing) chainDB
-
-        nrFetchClientRegistry = fetchClientRegistry
-
-        networkRequires :: NetworkRequires m up blk hdr
-        networkRequires = NetworkRequires {..}
+        responderApp :: NodeToNodeProtocols
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: This responder/initiator stuff doesn't really belong in `initInternalState`, it should move to `initNetworkLayer`.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 2]
```diff
@@ -464,34 +522,54 @@ forkBlockProduction IS{..} =
   New network layer
 -------------------------------------------------------------------------------}
 
-data NetworkRequires m up blk hdr = NetworkRequires {
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: I think we should revert this abstraction to its previous form so we can decouple creating the clients and servers from actually connecting them over channels.

File: ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs [hunk 3]
```diff
@@ -140,37 +193,70 @@ broadcastNetwork registry btime numCoreNodes pInfo initRNG numSlots = do
...
-                      => (   NodeChan m (SimpleBlock c ext) (SimpleHeader c ext)
-                          -> Channel m bytes
-                         )
-                      -> Codec ps e m bytes -> NodeComms m ps e bytes
-            mkCommsUp getChan codec = NodeComms {
-                ncCodec    = codec
-              , ncWithChan = \cc -> cc $
-                  loggingChannel (TalkingToProducer us them) $
-                    getChan (chans Map.! them Map.! us)
+          ni :: NetworkInterface m NodeId (AnyProtocolMessage (SimpleHeader c ext) (SimpleBlock c ext))
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs]: This is too much code, the flow is gone. Move it to some local or top-level helpers.

File: ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs [hunk 4]
```diff
@@ -209,18 +295,12 @@ broadcastNetwork registry btime numCoreNodes pInfo initRNG numSlots = do
...
 -------------------------------------------------------------------------------}
 
--- | Communication channel used for the Chain Sync protocol
-type ChainSyncChannel m hdr = Channel m (AnyMessage (ChainSync hdr (Point hdr)))
-
--- | Communication channel used for the Block Fetch protocol
-type BlockFetchChannel m blk = Channel m (AnyMessage (BlockFetch blk))
-
 -- | The communication channels from and to each node
 data NodeChan m blk hdr = NodeChan
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs]: Is it possible to revert this change and to use the "demuxed" channels here (I don't know whether that's possible).
Now, it says the the ChainSync consumer and producer can receive/send any message, including a BlockFetch message (and vice versa). Maybe this won't be the case at run-time, but it would be so much nicer to have this at compile-time too.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs (hunk 1 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: If we reinstate the `NetworkRequires` abstraction (see next comment), these `nr-` definitions can be brought back.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 5]
```diff
@@ -464,34 +522,54 @@ forkBlockProduction IS{..} =
...
--- | Required by the network layer to initiate comms to a new node
-data NodeComms m ps e bytes = NodeComms {
-      -- | Codec used for the protocol
-      ncCodec    :: Codec ps e m bytes
-
-      -- | Construct a channel to the node
+data NetworkRequires m up blk hdr bytes a b = NetworkRequires {
+      -- | Mux application which runs with upstream peers.
+      --
+      nrMuxInitiatorApp     :: up
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: I would expect this `Versions` stuff to be encapsulated. This is specific to a muxed communication channel.

**Comment by coot** [comment]
@mrBliss it is possible to have both decoupled and muxed channels even in the current form. And this is done in the tests: `test-consensus` is running without the mux layer.  I was rather thinking that  we should provide such an abstraction in `ouroboros-network`, and it could be based on the code I wrote for `test-consensus`. It's interface would be:
```
runWithoutMux :: NodeChanns m blk hdr -> ... -> NetworkInterface m NetworkAddress bytes
```

Another problem yet to solve is to to abstract over the handshake protocol: in tests it's not really needed, but the real version needs to run it.

File: ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs (hunk 4 above)
Comment (coot) [reply ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs]: This was not possible, for that reason in tests I needed to add `AnyProtocolMessage`.  The tests still run using a form of `Id` codec: still the only serialisation / de-serialisation that happens is wrapping and unwrapping of some existential types, but we loose type safety since the channels needed to have the same type.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs (hunk 5 above)
Comment (coot) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: I agree, if we abstract over it it will be easier to do different things: in this form handshake negotation is forced upon the node which I am not happy with.

**Comment by avieth** [comment]
> you have updated the abstractions to fit the new approach. But in doing so, you have hardcoded too many details from the network implementation, IMO.
> 
> It would be much better to update the abstraction in such a way that it can accommodate both approaches: mux all protocols over a single channel or use a channel for each protocol. That way, we could add another flag to the demo to determine how it should run: either using a single TCP connection or using named pipes. This would be a nice proof that our abstraction is abstract enough. Furthermore, this would minimise the changes needed in the Node module.

Agreed. In an ideal world each protocol would be driven independently with its own `Channel`, and we'd have something like...

```Haskell
-- Split a byte channel into many channels. Must be in IO continuation passing style becaue
-- the channels will be sourced by some thread  reading the input channel.
muxChannel :: MuxCodec ptcl -> Vec n ptcl -> Channel IO ByteString -> (Vec n (Channel IO ByteString) -> IO t) -> IO t
```

To mux, split the channel with `muxChannel` and run each protocol independently. To not mux, get all the channels you need from individual sockets.

I assume there's some technical reasons why something like `muxChannel` isn't feasible.

**Comment by coot** [comment]
> I assume there's some technical reasons why something like muxChannel isn't feasible.

The current `muxStart` is more like (mxu is usiing `(->) ptcl` rather than `Vect n`)
```
muxChannel ::  MuxBearer ptcl m -> (ptcl -> (Channel IO ByteString, Channel IO ByteString) -> IO t) -> IO t
```
The `MuxBearer` is adding it's own headers which are important for its operation, and it also gives back timing information (which will be important for measuring ΔQs): so it's not just `Channel m (MuxSDU ptcl)`.  If we move the time measure to the thread that runs the connection it will no longer measure how long it took to receive a message but it will also include how long it took for the mux layer to handle the message.

**Comment by coot** [comment]
An an after thought, we could remove the timing part of `MuxBearer` and measure time in `muxStart`, this  render the types of channels more aligned.

**Comment by coot** [comment]
After consulting with @semanticphilosopher, the timing information is essential and some bearers might gives us information available from the kernel when a packet arrived, thus we should not simplify this part.

**Comment by dcoutts** [comment]
> That way, we could add another flag to the demo to determine how it should run: either using a single TCP connection or using named pipes. This would be a nice proof that our abstraction is abstract enough. Furthermore, this would minimise the changes needed in the Node module.

@mrBliss mr It's not clear to me that this is an abstraction we need to support. We can easily handle the local case by using a single local unix domain sockets or windows named pipe, and using the mux & version negotiation over that. Supporting multiple local unix FIFOs is not something we need to do. The only plausible case I see here is wanting to run tests using identity codecs and no mux. But how important is this? We can still run in-memory tests with the mux, it's just easier to do that with the real codecs, and I don't see that including them is problematic.

**Comment by coot** [comment]
@dcoutts the consensus tests in this branch are running over identity codecs without multiplexing.

**Comment by avieth** [comment]
1. It would be really useful to have a way to easily switch off mux without having to do a big refactor. 
2. And conversely, how easy is it to get a mux channel _without_ typed protocols above it? To send/receive arbitrary bytes in arbitrary "protocol" numbers.

I'm interested in these things today in particular because I was debugging a space leak in byron-proxy's chain sync server, and so wanted to strip out as many features as possible in an effort to isolate the offending part of the program. I went ahead and got rid of mux by painstaking manual intervention, to find the leak disappeared!

If we could do item 1, I could have found this within minutes rather than hours. If we could do item 2, we could easily verify that mux is not leaking (or doing other buggy things).

**Comment by coot** [comment]
@avieth it's indeed not easy to switch off mux as of today, what I have here in the conensus test's would be really useful to you: here one gives a map of channels between peers, and receives back `connectTo` functions and accept loop for each peer, which runs without mux layer.  My aim is to provide:
```
muxLessConnect
  :: Map nodeId (Map nodeId (ptcl -> Channel))) -- ^ initial channels between peers
  -> Map nodeId (Map nodeId MuxApplication)
  -> ( nodeId 
      -> (nodeId -> m ()) -- connect to node id
     -> m ()                    -- accept loop
     -> m x)
  -> m x
```

**Comment by karknu** [comment]
>2. And conversely, how easy is it to get a mux channel without typed protocols above it? To send/receive arbitrary bytes in arbitrary "protocol" numbers.
I agree with both statement 1 and 2.
It is possible to run a mux channel as just a ByteStream channel without a typed protocol, this is done by this test in https://github.com/input-output-hk/ouroboros-network/blob/1c291da70decc003b474b192e335e86e80c252b3/ouroboros-network/test/Test/Mux.hs#L590

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 6]
```diff
@@ -456,122 +539,88 @@ forkBlockProduction IS{..} =
   New network layer
 -------------------------------------------------------------------------------}
 
-data NetworkRequires m up blk = NetworkRequires {
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: I liked the `NetworkRequires` record, it contained the server and clients of the protocol. Is there a reason for getting rid of it? Looking at `consensusNetworkApps`, I can't seem to find one.

**Comment by mrBliss** [comment]
@coot If you could maybe have a final go at reintroducing `NetworkRequires` (see my comment)? As I'll be gone next week, feel free to merge this when I'm gone.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs (hunk 6 above)
Comment (coot) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: I renamed it as `NetworkApplication` (this is the network requirement), this matches better with names in `ouroboros-network`.

**Comment by coot** [comment]
@mrBliss I just renamed it (maybe I should document it better), cheers.

**Comment by mrBliss** [comment]
> @mrBliss I just renamed it (maybe I should document it better), cheers.

I can't seem to find it.

This is what I'm talking about:
```haskell
data NetworkRequires m up blk = NetworkRequires {
      -- | Start a chain sync client that communicates with the given upstream
      -- node.
      nrChainSyncClient     :: up -> ChainSyncClient (Header blk) (Point blk) m Void

      -- | Start a chain sync server.
    , nrChainSyncServer     :: ChainSyncServer (Header blk) (Point blk) m ()

      -- | Start a block fetch client that communicates with the given
      -- upstream node.
    , nrBlockFetchClient    :: BlockFetchClient (Header blk) blk m ()

      -- | Start a block fetch server server.
    , nrBlockFetchServer    :: BlockFetchServer blk m ()

      -- | The fetch client registry, used by the block fetch client.
    , nrFetchClientRegistry :: FetchClientRegistry up (Header blk) blk m
    }
```

**Comment by coot** [comment]
It evolved into:
```
data NetworkApplication m versions up blk bytes a b = NetworkApplication {
      -- | Mux application which runs with upstream peers.
      --
      naMuxInitiatorApp :: up
                        -> versions
                            (MuxApplication
                              InitiatorApp
                              NodeToNodeProtocols
                              m bytes (a, Void))

      -- | Mux application which serves downstream peers.
      --
      -- TODO: use 'AnyMuxResponderApp', but it requires better connection
      -- management on the networking side.  This way we avoid simultaneously
      -- opened connections for now.
      --
    , naMuxResponderApp :: versions
                            (MuxApplication
                              ResponderApp
                              NodeToNodeProtocols
                              m bytes (Void, b))
    }
```

Where `MuxApplication` is:
```
data MuxApplication (appType :: AppType) ptcl m bytes a where
  MuxInitiatorApplication
    -- Initiator application; most simple application will be @'runPeer'@ or
    -- @'runPipelinedPeer'@ supplied with a codec and a @'Peer'@ for each
    -- @ptcl@.  But it allows to handle resources if just application of
    -- @'runPeer'@ is not enough.  It will be run as @'ModeInitiator'@.
    :: m (ptcl -> Channel m bytes ->  m a)
    -> MuxApplication InitiatorApp ptcl m bytes (a, Void)

  MuxResponderApplication
    -- Responder application; similarly to the @'MuxInitiatorApplication'@ but it
    -- will be run using @'ModeResponder'@.
    :: m (ptcl -> Channel m bytes ->  m a)
    -> MuxApplication ResponderApp ptcl m bytes (Void, a)

  MuxInitiatorAndResponderApplication
    -- Initiator and server applications.
    :: m ( ptcl -> Channel m bytes ->  m a
         , ptcl -> Channel m bytes ->  m b
         )
    -> MuxApplication InitiatorAndResponderApp ptcl m bytes (a, b)
```

**Comment by mrBliss** [comment]
@coot Ah, I see. That is unrecognisable :slightly_frowning_face:

**Comment by coot** [comment]
@mrBliss hence a new name ;).  In the `ouroboros-network` we are using `ptcl -> Channel m ByteString -> m ()`, where `ptcl` is an `Enum` type.  This way the network layer  can handle any number of mini-protocols.  Here in consensus layer `ptcl` is instantiated to `NodeToNodeProtocol` which (now just contains two constructors) this way it makes it much more similar to the original record.

File: ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs [hunk 7]
```diff
@@ -99,62 +98,71 @@ type family HasResponder (appType :: AppType) :: Bool where
...
+data MuxApplication (appType :: AppType) ptcl m bytes a where
   MuxInitiatorApplication
     -- Initiator application; most simple application will be @'runPeer'@ or
     -- @'runPipelinedPeer'@ supplied with a codec and a @'Peer'@ for each
     -- @ptcl@.  But it allows to handle resources if just application of
     -- @'runPeer'@ is not enough.  It will be run as @'ModeInitiator'@.
-    :: (ptcl -> Channel m ByteString ->  m ())
-    -> MuxApplication InitiatorApp ptcl m
+    :: m (ptcl -> Channel m bytes ->  m a)
+    -> MuxApplication InitiatorApp ptcl m bytes (a, Void)
```
Comment (avieth) [inline ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs]: The use of `(a, Void)` here requires a comment. Why not instead have 2 parameters: one for the initiator result type, one for that of the responder?

File: ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs [hunk 8]
```diff
@@ -99,62 +98,71 @@ type family HasResponder (appType :: AppType) :: Bool where
...
+--
+data MuxApplication (appType :: AppType) ptcl m bytes a where
   MuxInitiatorApplication
     -- Initiator application; most simple application will be @'runPeer'@ or
     -- @'runPipelinedPeer'@ supplied with a codec and a @'Peer'@ for each
     -- @ptcl@.  But it allows to handle resources if just application of
     -- @'runPeer'@ is not enough.  It will be run as @'ModeInitiator'@.
-    :: (ptcl -> Channel m ByteString ->  m ())
-    -> MuxApplication InitiatorApp ptcl m
+    :: m (ptcl -> Channel m bytes ->  m a)
```
Comment (avieth) [inline ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs]: What's the motivation for this change? I remember arguing against a similar thing in time-warp ages ago.

Comment (avieth) [reply ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs]: (referring to the addition of `m` out front, not the new type parameter)

File: ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs [hunk 9]
```diff
@@ -98,46 +98,55 @@ type family HasResponder (appType :: AppType) :: Bool where
 --   serving downstream peers using server side of each protocol and getting
 --   updates from upstream peers using client side of each of the protocols.
 --
+-- * the monadic action which wraps the callbacks will be run only once by the
+--   multiplexing layer.  This makes it safe for sharing a common mutable state
+--   between various protocol.
```
Comment (avieth) [inline ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs]: When will it be run? Why can't the programmer run it before setting up the mux/server stuff?

Comment (avieth) [reply ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs]: Clarification: it will be run only once _per peer_. Correct?

Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs]: Indeed, once per connection. Probably much better type is this one:
```
data MuxApplication (appType :: AppType) ptcl m bytes a where
  MuxInitiaroApplication
    :: forall s appType ptcl m bytes a.
        m s
    -> (s -> ptcl -> Channel m bytes -> m a)
    -> MuxApplication appType ptcl m bytes a
...
```
What I need is a shared state between all the `ptcl` threads, which is constructed before starting all the mux threads.

Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs]: This is needed because there's synchronisation between chain sync and block fetch clients in the consensus layer  (chain sync client has to start after block fetch).

Comment (avieth) [reply ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs]: > chain sync client has to start after block fetch

What's the reason for it? Seems like something that ought to be fixed.

Comment (avieth) [reply ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs]: I'd imagine (admitting that I know very little of the actual implementation) that block fetch and chain sync parts would synchronize by way of some shared STM state

Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs]: It's explained here: https://github.com/input-output-hk/ouroboros-network/blob/f3d6569bbed21628252c6200f6c63309acb886e4/ouroboros-consensus/src/Ouroboros/Consensus/Node.hs#L300

File: ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs (hunk 7 above)
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs]: You mean `MuxApplication appType ptcl m a b`, right this could be fixed.  I tried this at first but I run into some problems, but I don't see why in the first place.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 10]
```diff
@@ -484,12 +483,16 @@ forkBlockProduction IS{..} =
   New network layer
 -------------------------------------------------------------------------------}
 
+data SharedState m s a = SharedState (m s) (s -> a)
```
Comment (avieth) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: I still argue that this whole thing should not be necessary. The fix is in the wrong place. The block fetch and chain sync client _should not_ require an ordering on their startup. @dcoutts @coot

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs [hunk 11]
```diff
@@ -49,148 +52,184 @@ import           Ouroboros.Consensus.ChainSyncServer
...
-      -- itself to register and deregister peers.p
-    , addDownstream :: forall eCS eBF bytesCS bytesBF.
-                       (Exception eCS, Exception eBF)
-                    => NodeComms m (ChainSync (Header blk) (Point blk)) eCS bytesCS
-                       -- Communication for the Chain Sync protocol
-                    -> NodeComms m (BlockFetch blk)                     eBF bytesBF
-                       -- Communication for the Block Fetch protocol
-                    -> m ()
-    }
+data SharedState m s a = SharedState (m s) (s -> a)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs]: Just like @avieth, I'm not a fan of this `SharedState`. I realise why it is needed (I added the `clientRegistered` `TMVar` stuff in the first place to appease the invariants in the block fetch client), but it would be nicer if we could solve it in another way.

Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs]: BTW, why is the `Semigroup` instance needed again? After removing it, the demo and the consensus tests still compile.

Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs]: Just thinking out loud:

Does `SharedState` have to be so general? In particular, does it have to parameterise over `s`? Typically, I like to be as general as possible too, but in this case, being more specific might be clearer?

For example, can we rename it to `InitNetwork(Apps) m a` and fix `s` to `TMVar m ()`?

Alternatively, could it maybe take the `TMVar` as an argument?

**Comment by mrBliss** [comment]
BTW, the demo is still not shutting down properly when terminating it with <kbd>CTRL</kbd>-<kbd>C</kbd>, it's giving an assertion failure.

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs (hunk 11 above)
Comment (coot) [reply ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs]: I probably wasn't clear; `Semigroup` instance would be useful to create multiple versions.  In the demo we use `SharedState`'s functor instance to map `SharedState (TMVar m s) (NetworkApplication ...)` into `SharedState (TMVar m s) (Versions ....)`. `Versions` is basically a `Map versionNumber (NetworkApplication ...)` (it's slightly more general than that but for this context it's enough to think of it as that). Both sides will agree on a single version to run.  To build a non singleton we need a way to combine multiple `SharedState m s (Versions ...)` into one, that's how semigroup instance could be used.  Right now, we have just a single protocol version so the semigroup instance is not really needed.

Comment (coot) [reply ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs]: I will rename it and I'll fix the `s` variable.

File: ouroboros-consensus/demo-playground/Run.hs [hunk 12]
```diff
@@ -78,7 +87,6 @@ handleSimpleNode p CLI{..} (TopologyInfo myNodeId topologyFile) = do
 
     putStrLn $ "**************************************"
     putStrLn $ "I am Node = " <> show myNodeId
-    putStrLn $ "My consumers are " <> show (consumers nodeSetup)
```
[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-consensus/demo-playground/Run.hs]: Because consumers will connect to us, when we start the node we don't know who will connect to us. `NodeSetup` does not contain `consumers` any more.

The information is not lost though, it is still in `NetworkTopology ~ [NodeSetup]` but one would need to go through `NodeSetup` of all the other nodes to see who will contact us; I just don't think this would be the right thing to do.

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs [hunk 13]
```diff
@@ -7,36 +7,39 @@
 {-# LANGUAGE RankNTypes                 #-}
 {-# LANGUAGE RecordWildCards            #-}
 {-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TypeApplications           #-}
 {-# LANGUAGE TypeFamilies               #-}
```
[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs]: Sure, I'll remove it.

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs [hunk 14]
```diff
@@ -49,148 +52,184 @@ import           Ouroboros.Consensus.ChainSyncServer
...
+    -- this is not ideal.  Currently mux layer does not support custom monadic
+    -- expression before it starts; we either need to fix the client side of
+    -- block fetch & chain sync protocols or generalise mux.  SharedState here
+    -- makes additional problem: to build version dictionary
+    -- from singletons we need to use the left biased semigroup instance of
+    -- 'SharedState'.
+    -> SharedState m (TMVar m ())
+        (NetworkApplication m peer bytesCS bytesBF ())
+
+consensusNetworkApps traceChainSync traceBlockFetch codecChainSync codecBlockFetch NodeParams {..} kernel =
```
[the remark this answers is not in the document]
Comment (dcoutts) [reply ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs]: How about we expose more stuff through the `NodeKernel` and don't pass `NodeParams`.

File: ouroboros-network/demo/chain-sync.hs [hunk 15]
```diff
@@ -388,6 +388,10 @@ clientBlockFetch sockAddrs = do
         muxApplication peerid =
           MuxInitiatorApplication (protocols peerid)
 
+        prunePeerSTM :: FilePath -> STM ()
```
Comment (mrBliss) [inline ouroboros-network/demo/chain-sync.hs]: `FilePath`?

Comment (coot) [reply ouroboros-network/demo/chain-sync.hs]: Surprising isn't it, it was forced by the caller, but we don't need to worry about this, this patch was deleted.

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs [hunk 16]
```diff
@@ -49,142 +61,237 @@ import           Ouroboros.Consensus.ChainSyncServer
...
+data ProtocolHandlers m peer blk = ProtocolHandlers {
+      phChainSyncClient
+        :: peer
+        -> ChainSyncClient (Header blk) (Point blk) m Void
+
+    , phChainSyncServer
+        :: ChainSyncServer (Header blk) (Point blk) m ()
 
+    -- TODO block fetch client does not have GADT view of the handlers.
+    , phBlockFetchClient
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs]: Can you use the `BlockFetchClient` type synonym from https://github.com/input-output-hk/ouroboros-network/blob/2529bf78a5f8c6900fc74338473cea9785515eb7/ouroboros-network/src/Ouroboros/Network/BlockFetch/Client.hs#L58

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs [hunk 17]
```diff
@@ -49,142 +61,237 @@ import           Ouroboros.Consensus.ChainSyncServer
...
-      -- This is in CPS style to allow for resource allocation.
-    , ncWithChan :: forall a. (Channel m bytes -> m a) -> m a
-    }
+runSharedState :: Monad m => SharedState m s a -> m a
+runSharedState (SharedState ms f) = ms >>= pure . f
+
+
+-- | Protocol handlers as provided by the consensus.
+--
+data ProtocolHandlers m peer blk = ProtocolHandlers {
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs]: Ah, finally I get my `NetworkRequires` record back, but with a better name :slightly_smiling_face:

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs [hunk 18]
```diff
@@ -49,142 +61,237 @@ import           Ouroboros.Consensus.ChainSyncServer
...
+-- | Protocol codecs needed to run 'ProtocolHandlers'.
+--
+data ProtocolCodecs blk failure m bytesCS bytesBF = ProtocolCodecs {
+    pcChainSyncCodec  :: Codec (ChainSync (Header blk) (Point blk)) failure m bytesCS
+  , pcBlockFetchCodec :: Codec (BlockFetch blk) failure m bytesBF
+  }
+
+-- | Id codecs used in tests.
+--
+protocolCodecsId :: Monad m => ProtocolCodecs blk CodecFailure m (AnyMessage (ChainSync (Header blk) (Point blk))) (AnyMessage (BlockFetch blk))
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs]: Maybe wrap this line like so:
```haskell
protocolCodecsId :: Monad m
                 => ProtocolCodecs blk CodecFailure m (AnyMessage (ChainSync (Header blk) (Point blk)))
                                                      (AnyMessage (BlockFetch blk))
protocolCodecsId = ..
```

File: ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs [hunk 19]
```diff
@@ -58,6 +63,88 @@ import qualified Ouroboros.Storage.ChainDB.API as ChainDB
...
+        (NetworkApplication m nodeId
+          (AnyMessage (ChainSync (Header blk) (Point blk)))
+          (AnyMessage (BlockFetch blk)) ())
+    -> NetworkInterface m nodeId
+createNetworkInterface chans nodeIds us na = NetworkInterface
+    { niConnectTo = \them -> do
+        NetworkApplication {naChainSyncClient, naBlockFetchClient} <- runSharedState na
+        let nodeChan = chans Map.! them Map.! us
+
+        -- 'withAsync' guarantess that when 'waitAny' below receives an
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs]: `s/guarantess/guarantees/`, `s/kills/killed/`

File: ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs [hunk 20]
```diff
@@ -58,6 +63,88 @@ import qualified Ouroboros.Storage.ChainDB.API as ChainDB
...
+                  -- of them; this is consistent with
+                  -- 'Ouroboros.Network.Socket.connectTo'.
+                  void $ waitAny [aCS, aBF]
+
+      , niWithServerNode = \k -> mask $ \unmask -> do
+        ts :: [Async m ()] <- fmap concat $ forM (filter (/= us) nodeIds) $ \them -> do
+              NetworkApplication {naChainSyncServer, naBlockFetchServer} <- runSharedState na
+              let nodeChan = chans Map.! us Map.! them
+
+              aCS <- async $ unmask
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs]: Why not use `withAsync` here too?

File: ouroboros-consensus/demo-playground/Run.hs [hunk 21]
```diff
@@ -130,15 +140,66 @@ handleSimpleNode p CLI{..} (TopologyInfo myNodeId topologyFile) = do
...
+                addr:_ <- getAddrInfo Nothing (Just host) (Just port)
+                versions <- runSharedState networkApp
+                connectTo
+                      (muxInitiatorNetworkApplication na <$> versions)
+                      -- Do not bind to a local port, use ephemeral
+                      -- one.  We cannot bind to port on which the server is
+                      -- already accepting connections.
+                      Nothing
+                      addr
+                  `catch` \(_ :: IOException) -> threadDelay 250_000 >> io
```
Comment (dcoutts) [inline ouroboros-consensus/demo-playground/Run.hs]: Lets add a TODO here to note that this delay loop will be replaced by the connection logic later.

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs (hunk 17 above)
Comment (coot) [reply ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs]: :)
